File: 42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <=2: return 0
        i, j = 1, len(height) - 1
        lmax, rmax = height[0], height[-1]
        res = 0
        while i <= j:
            if height[i] > lmax: lmax = height[i]
            if height[j] > rmax: rmax = height[j]
            
            if lmax <= rmax:
                res += lmax - height[i]
                i += 1
            else:
                res += rmax - height[j]
                j -= 1
        return res

File: 42. Trapping Rain Water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if(height.size() <= 2) return 0;
        int l = 0, r = height.size() - 1;
        int lmax = height[l], rmax = height[r];
        int res = 0;
        while(l < r){
            if(lmax < height[l]){lmax = height[l];}
            if(rmax < height[r]){rmax = height[r];}
            
            if(lmax < rmax){
                res += lmax - height[l];
                l++;
            }
            else{
                res += rmax - height[r];
                r--;
            }
        }
        return res;
    }
};
